src/number_theory/modular_forms/congruence_subgroups.lean

```diff
@@ -0,0 +1,234 @@
+right-hand entry. -/
+def Gamma_0_map (N : ℕ): (Gamma0 N) →* (zmod N) :=
+{ to_fun := λ g, ((↑ₘg 1 1 : ℤ) : zmod N),
+  map_one' := by { simp, },
+  map_mul' := by {intros A B,
+  have := (mat_two_mul_expl A.1.1 B.1.1).2.2.2,
+  simp only [coe_coe, subgroup.coe_mul, coe_matrix_coe, coe_mul, int.coe_cast_ring_hom, map_apply,
+    subtype.val_eq_coe, mul_eq_mul] at *,
+  rw this,
+  have ha := A.property,
+  simp only [int.cast_add, int.cast_mul, add_left_eq_self, subtype.val_eq_coe, Gamma0_mem, coe_coe,
+    coe_matrix_coe, int.coe_cast_ring_hom, map_apply] at *,
+  rw ha,
+  simp,} }
+
+/--The congruence subgroup `Gamma1` (as a subgroup of `Gamma0`) of matrices whose bottom
+row is congruent to `(0,1)` modulo `N`.-/
+def Gamma1' (N : ℕ) : subgroup (Gamma0 N) := (Gamma_0_map N).ker
+
+@[simp]
+lemma Gamma1_mem' (N : ℕ) (γ :(Gamma0 N)) : γ ∈ (Gamma1' N) ↔ ((Gamma_0_map N) γ) = 1 := iff.rfl
+
+lemma Gamma1_to_Gamma0_mem (N : ℕ) (A : Gamma0 N) : A ∈ (Gamma1' N) ↔
+  ((↑ₘA 0 0 : ℤ) : zmod N) = 1 ∧ ((↑ₘA 1 1 : ℤ) : zmod N) = 1 ∧ ((↑ₘA 1 0 : ℤ) : zmod N) = 0 :=
+begin
+  split,
+  intro ha,
+  have hA := A.property,
+  rw Gamma0_mem at hA,
+  have adet := Gamma0_det N A,
+  rw matrix.det_fin_two at adet,
+  simp only [Gamma_0_map, coe_coe, coe_matrix_coe, int.coe_cast_ring_hom, map_apply, Gamma1_mem',
+    monoid_hom.coe_mk, subtype.val_eq_coe, int.cast_sub, int.cast_mul] at *,
+  rw [hA, ha] at adet,
+  simp only [mul_one, mul_zero, sub_zero] at adet,
+  simp only [adet, hA, ha, eq_self_iff_true, and_self],
+  intro ha,
+  simp only [Gamma1_mem', Gamma_0_map, monoid_hom.coe_mk, coe_coe, coe_matrix_coe,
+    int.coe_cast_ring_hom, map_apply],
+  exact ha.2.1,
+end
+
+/--The congruence subgroup `Gamma1` of `SL(2,ℤ)` consisting of matrices whose bottom
+row is congruent to `(0,1)` modulo `N`. -/
+def Gamma1 (N : ℕ) : subgroup SL(2, ℤ) := subgroup.map
+(((Gamma0 N).subtype).comp (Gamma1' N).subtype) ⊤
+
+@[simp]
+lemma Gamma1_mem (N : ℕ) (A : SL(2, ℤ)) : A ∈ (Gamma1 N) ↔
+  ((↑ₘA 0 0 : ℤ) : zmod N) = 1 ∧ ((↑ₘA 1 1 : ℤ) : zmod N) = 1 ∧ ((↑ₘA 1 0 : ℤ) : zmod N) = 0 :=
+begin
+  split,
+  { intro ha,
+    simp_rw [Gamma1, subgroup.mem_map] at ha,
+    simp at ha,
+    obtain ⟨⟨x, hx⟩, hxx⟩ := ha,
+    rw Gamma1_to_Gamma0_mem at hx,
+    rw ←hxx,
+    convert hx },
+  { intro ha,
+    simp_rw [Gamma1, subgroup.mem_map],
+    have hA : A ∈ (Gamma0 N), by {simp [ha.right.right, Gamma0_mem, subtype.val_eq_coe],},
+    have HA : (⟨A , hA⟩ : Gamma0 N) ∈ Gamma1' N,
+      by {simp only [Gamma1_to_Gamma0_mem, subgroup.coe_mk, coe_coe, coe_matrix_coe,
+        int.coe_cast_ring_hom, map_apply],
+      exact ha,},
+    refine ⟨(⟨(⟨A , hA⟩ : Gamma0 N), HA ⟩ : (( Gamma1' N ) : subgroup (Gamma0 N))), _⟩,
+    simp }
+end
+
+lemma Gamma1_in_Gamma0 (N : ℕ) : (Gamma1 N) ≤ (Gamma0 N) :=
+begin
+  intros x HA,
+  simp only [Gamma0_mem, Gamma1_mem, coe_coe, coe_matrix_coe, int.coe_cast_ring_hom,
```

Comment:
Is it possible to close the goal using `simpa`?

Reply:
I couldn't seem to get it to work, but maybe I'm not doing the `simpa` correctly.